MISRA-C Use gcc flags to enforce some rules

ceolin opened the pull request:
GCC contains warning flags that can partially help spot MISRA-C violations. This pr is enforcing checks for rules:

8.2, 8.3, 8.4, 16.1 and 16.3

Commit:
kernel: timeout: Remove unused parameter

Function z_init_timeout was receivig two parameters, a timeout and a
callback function. The problem was that the callback function was not
being used, this callback should be set in z_add_timeout function.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
drivers: timer: Explicitly setting argument as unused

Marking as unused (ARG_UNUSED) the parameter device in the
initialization function z_clock_driver_init when it is not used.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
compiler: gcc: Add warning option to spot poblematic switches

Option -Wimplicit-fallthrough as the name suggests raises a warning when
there is an implicit fallthrough in a switch clause. This warning does
not happen when the last statement is a return or a call to a function
declared with noreturn attribute.

This flag helps to enforce two MISRA-C rules (not completely), 16.1 and
16.3

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
compiler: gcc: Add warning when overriding and object initialization

The option -Woverride-init warns when an element of an object is
initialized more than once what is usually a programming mistake. e.g

int x[3] = {[0] = 1, [1] = 2, [1] = 2}

Helps to enforce MISRA-C rule 9.4

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
kernel: poll: Avoid initialize same object twice

When CONFIG_POLL is enabled, _K_QUEUE_INITIALIZER was initializing the
same enum twice. This is a problem because order in which any side
effects occur among the initialization list is unspecified.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
gcc: Replace passthrough with fallthrough

Fallthrough is a string recognized by gcc in comments that tells gcc
that a specific fallthrough in switch/case statement is intentional.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
net: ip: Changing comment to inform gcc about fallthrough

-Wimplicit-fallthrough=2 raises a warn when there is a switch clause
 without a break statement unless it finds a fallthrough comment.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
bluetooth: Make explicit fallthrough cases

-Wimplicit-fallthrough=2 requires a fallthrough comment to tells gcc
 that this happens intentionally.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
net: dns: Disable implicit-fallthrough option

gcc has problems with comments for clauses under ifdef.
Just disabling this check for a specific switch statement to avoid
wrong warning messages.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
net: http: Changing comment to inform gcc about fallthrough

-Wimplicit-fallthrough=2 raises a warn when there is a switch clause
 without a break statement unless it finds a fallthrough comment.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
usb: Make explicit fallthrough cases

-Wimplicit-fallthrough=2 requires a fallthrough comment to tells gcc
 that this happens intentionally.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
misra: Use gcc flags to sanityze function declaration

Force functions to have named parameters and their types and a prototype
for functions with external linkage.

MISRA-C related rules 8.2, 8.3 and 8.4

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
gcc: Add warning for empty if, else and do while body

Warn if an empty body occurs in an if, else or do while statement.
e.g: if(x); else; ...

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

andyross reviewed (approved):
I had no idea that gcc would disable that switch warning if you put "fallthrough" in a comment.  Seems crazy, but not wrong.  I guess we have to go back and change all those spots, though, right?

ceolin commented:
I don't know how to deal with eternal modules. It seems they are using same flags used across the build. We may need turn some  flags off for them

Commit:
bluetooth: Disable implicit-fallthrough option

gcc has problems with comments for clauses under ifdef.
Just disabling this check for a specific switch statement to avoid
wrong warning messages.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

pfalcon commented:
> @pfalcon pfalcon removed request for pfalcon and nategraff-sifive now

I actually removed just myself, Github did the rest (due to 15 reviewers limit; heh, if 640K should be enough for everyone, then 15 reviewers should ne for suremore, d'oh). So, @nategraff-sifive, if you're interested in reviewing this, please do.

Commit:
disk: Make explicit fallthrough cases

-Wimplicit-fallthrough=2 requires a fallthrough comment to tells gcc
 that this happens intentionally.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

ceolin commented:
https://github.com/apache/mynewt-mcumgr/pull/29

Commit:
i2c: Make explicit fallthrough cases

-Wimplicit-fallthrough=2 requires a fallthrough comment to tells gcc
 that this happens intentionally.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
toolchain: gcc: Macro to ignore implicit-fallthrough

Old GCC versions don't have implicit-fallthrough checks. This flag is
not being set when using these compilers. The problem is that we need
to explicitly disable this warning when the fallthrough comment is
inside an ifdfef clause. This was causing problems with these old
compilers.

Just abstracting it to avoid this problem.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

ceolin commented:
shippable is failing because need this pr https://github.com/zephyrproject-rtos/mcumgr/pull/4

dleach02 commented:
@ceolin, do we want to continue to pursue this PR?

joerchan commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -473,6 +473,12 @@ void ll_rx_dequeue(void)
 
 	mem_release(link, &mem_link_rx.free);
 
+	/* GCC is not able to recognize comments in clause under ifdef.
```
This could be fixed by moving the fallthrough comment out of the ifdef.

If not then I think those comments should be removed. Because the are obvious, and they don't work.

joerchan reviewed (approved):
Approving the Bluetooth part. One nit-pick comment but not important.

dleach02 commented:
First, at a minimum, if we are going to use comments then I suggest using a common phrase. In this PR there is "fall through", "falls through", and "fallthrough" used with at least one file using two different styles. I get that the "-Wimplicit-fallthrough=2" is supposed to match some of these but for consistency we should use the same value.

Second, not sure if this was considered, but using gcc  "__attribute__((fallthrough))" to give a compiler information that the "-Wimplicit-fallthrough" warning should not be emitted. My suggestion would be to create a macro for this attribute that would provide different fillers based on the compiler. This would still give the reader a clear indication that the developer intended the case logic to fall through to the next case.

gcc: #define ZEPHYR_FALLTHROUGH  __attribute__ ((fallthrough))
c++11/c++14: #define ZEPHYR_FALLTHROUGH [[gnu::fallthrough]]
clang: #define ZEPHYR_FALLTHROUGH [[clang::fallthrough]]
c++17 and above: #define ZEPHYR_FALLTHROUGH [[fallthrough]] 

MBED has a variation of this in their platform/mbed_toolchain.h

pabigot commented:
> First, at a minimum, if we are going to use comments then I suggest using a common phrase. In this PR there is "fall through", "falls through", and "fallthrough" used with at least one file using two different styles. I get that the "-Wimplicit-fallthrough=2" is supposed to match some of these but for consistency we should use the same value.

General practice has been `/* fallthrough */` in C.  Linux used to use this, but currently defines a pseudo-keyword macro so it's `fallthrough;`.  I like `[[fallthrough]]` but that won't happen until Zephyr grows up to handle C18.

See [Linux current practice](https://www.kernel.org/doc/html/latest/process/deprecated.html#implicit-switch-case-fall-through) for details.  Zephyrs `checkpatch.pl` has removed the check that's in upstream because we don't have the pseudo-keyword.

ceolin commented:
> > First, at a minimum, if we are going to use comments then I suggest using a common phrase. In this PR there is "fall through", "falls through", and "fallthrough" used with at least one file using two different styles. I get that the "-Wimplicit-fallthrough=2" is supposed to match some of these but for consistency we should use the same value.
> 
> General practice has been `/* fallthrough */` in C. Linux used to use this, but currently defines a pseudo-keyword macro so it's `fallthrough;`. I like `[[fallthrough]]` but that won't happen until Zephyr grows up to handle C18.
> 
> See [Linux current practice](https://www.kernel.org/doc/html/latest/process/deprecated.html#implicit-switch-case-fall-through) for details. Zephyr's `checkpatch.pl` has removed the check that's in upstream because we don't have the pseudo-keyword.

I've created the suggested fallthrough macro, but have done it in another pr, since this one contains other flags as well. will split them in different pr to make it easier to review. Please take  a look in https://github.com/zephyrproject-rtos/zephyr/pull/27752
I followed other similar macro names, e.g __packed to name it __falthrough

ceolin commented:
Most fixes were addressed in other prs and some of them are not necessary anymore. Just closing it.
